File: android-lib/app/src/main/java/dev/sontx/logme/Logger.java
package dev.sontx.logme;

import android.annotation.SuppressLint;
import android.util.Log;

import java.text.SimpleDateFormat;
import java.util.Date;

import dev.sontx.logme.uce.UCEHandler;

public final class Logger {
    private static final String TAG = Logger.class.getName();
    @SuppressLint("StaticFieldLeak")
    private static LogMe logMe;

    private static final SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss,SSS");

    static void setLogMe(LogMe logMe) {
        Logger.logMe = logMe;
    }

    public static void i(String tag, String message) {
        Log.i(tag, message);
        sendToSupervisor("INFO", message);
    }

    public static void d(String tag, String message) {
        Log.d(tag, message);
        sendToSupervisor("DEBUG", message);
    }

    public static void e(String tag, String message, Throwable throwable) {
        Log.e(tag, message, throwable);
        String stackTrace = "";
        if (throwable != null)
            stackTrace = "\n" + UCEHandler.getStackTrace(throwable);
        sendToSupervisor("ERROR", message + stackTrace);
    }

    private static void sendToSupervisor(String type, String message) {
        LogMe logMe = Logger.logMe;
        if (logMe != null) {
            String now = simpleDateFormat.format(new Date());
            logMe.sendLog(String.format("%s %s: %s", now, type, message));
        }
    }

    public static void i(String message) {
        i(TAG, message);
    }

    public static void d(String message) {
        d(TAG, message);
    }

    public static void e(String message, Throwable throwable) {
        e(TAG, message, throwable);
    }
}

File: DotNetLib/LogMe/Logger.cs
using System;
using System.Diagnostics;

namespace LogMe
{
    public sealed class Logger
    {
        private static LogMe logMe;

        internal static void SetLogMe(LogMe logMe)
        {
            Logger.logMe = logMe;
        }

        private static string NormalizeMessage(string type, string message, string tag = null)
        {
            return string.IsNullOrEmpty(tag)
                ? $"{type}: {message}"
                : $"{type}: [{tag}] {message}";
        }

        public static void I(string tag, string message)
        {
            var normalizedMessage = NormalizeMessage("INFO", message, tag);
            Trace.TraceInformation(normalizedMessage);
            SendToSupervisor(normalizedMessage);
        }

        public static void D(string tag, string message)
        {
            var normalizedMessage = NormalizeMessage("DEBUG", message, tag);
            Trace.WriteLine(normalizedMessage);
            SendToSupervisor(normalizedMessage);
        }

        public static void E(string tag, string message, Exception exception)
        {
            var normalizedMessage = NormalizeMessage("ERROR", message, tag);
            if (exception != null)
            {
                normalizedMessage = $"{normalizedMessage}{Environment.NewLine}{exception}";
            }
            Trace.TraceError(normalizedMessage);
            SendToSupervisor(normalizedMessage);
        }

        private static async void SendToSupervisor(string message)
        {
            var logMe = Logger.logMe;
            if (logMe != null)
            {
                var now = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss,fff}";
                await logMe.SendAsync($"{now} {message}");
            }
        }

        public static void I(string message)
        {
            I(null, message);
        }

        public static void D(string message)
        {
            D(null, message);
        }

        public static void E(string message, Exception exception = null)
        {
            E(null, message, exception);
        }
    }
}